File: src/main/java/li/klass/fhem/service/device/GraphDefinitionsForDeviceService.java
package li.klass.fhem.service.device;

import android.content.Context;
import android.support.annotation.NonNull;

import com.google.common.base.Function;
import com.google.common.base.Optional;
import com.google.common.base.Predicate;
import com.google.common.collect.ImmutableSet;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import li.klass.fhem.domain.log.ConcernsDevicePredicate;
import li.klass.fhem.service.graph.gplot.GPlotDefinition;
import li.klass.fhem.service.graph.gplot.GPlotHolder;
import li.klass.fhem.service.graph.gplot.SvgGraphDefinition;
import li.klass.fhem.service.room.RoomListService;
import li.klass.fhem.service.room.xmllist.XmlListDevice;

import static com.google.common.collect.FluentIterable.from;
import static li.klass.fhem.service.room.xmllist.XmlListDevice.hasName;

public class GraphDefinitionsForDeviceService {

    private final RoomListService roomListService;
    private final GPlotHolder gPlotHolder;

    private static final Logger LOGGER = LoggerFactory.getLogger(GraphDefinitionsForDeviceService.class);

    @Inject
    GraphDefinitionsForDeviceService(RoomListService roomListService, GPlotHolder gPlotHolder) {
        this.roomListService = roomListService;
        this.gPlotHolder = gPlotHolder;
    }

    public ImmutableSet<SvgGraphDefinition> graphDefinitionsFor(Context context, XmlListDevice device, Optional<String> connectionId) {
        final ImmutableSet<XmlListDevice> allDevices = roomListService.getAllRoomsDeviceList(connectionId, context)
                .getAllDevicesAsXmllistDevice();

        LOGGER.info("graphDefinitionsFor(name={},connection={})", device.getName(), connectionId.or("--"));
        ImmutableSet<SvgGraphDefinition> devices = from(allDevices)
                .filter(XmlListDevice.hasType("SVG"))
                .filter(hasConcerningLogDevice(allDevices, device))
                .filter(gplotDefinitionExists(allDevices))
                .transform(toGraphDefinition(allDevices)).toSet();
        for (SvgGraphDefinition svgGraphDefinition : devices) {
            LOGGER.info("graphDefinitionsFor(name={},connection={}) - found SVG with name {}", device.getName(), connectionId.or("--"), svgGraphDefinition.getName());
        }
        return devices;
    }

    @NonNull
    private Function<XmlListDevice, SvgGraphDefinition> toGraphDefinition(final ImmutableSet<XmlListDevice> allDevices) {
        return new Function<XmlListDevice, SvgGraphDefinition>() {
            @Override
            public SvgGraphDefinition apply(XmlListDevice svgDevice) {
                String logDeviceName = svgDevice.getInternal("LOGDEVICE").get();
                String gplotFileName = svgDevice.getInternal("GPLOTFILE").get();
                GPlotDefinition gPlotDefinition = gPlotHolder.definitionFor(gplotFileName, isConfigDb(allDevices)).get();

                List<String> labels = Arrays.asList(svgDevice.getAttribute("label")
                        .or("").replaceAll("\"", "").split(","));
                String title = svgDevice.getAttribute("title").or("");
                List<String> plotfunction = Arrays.asList(svgDevice.getAttribute("plotfunction").or("").trim().split(" "));

                return new SvgGraphDefinition(svgDevice.getName(), gPlotDefinition, logDeviceName, labels, title, plotfunction);
            }
        };
    }

    @NonNull
    private Predicate<XmlListDevice> gplotDefinitionExists(final ImmutableSet<XmlListDevice> allDevices) {
        return new Predicate<XmlListDevice>() {
            @Override
            public boolean apply(XmlListDevice input) {
                if (input == null) {
                    return false;
                }
                Optional<String> gplotFileName = input.getInternal("GPLOTFILE");
                return gplotFileName.isPresent()
                        && gPlotHolder.definitionFor(gplotFileName.get(), isConfigDb(allDevices)).isPresent();
            }
        };
    }

    @NonNull
    private Predicate<XmlListDevice> hasConcerningLogDevice(final ImmutableSet<XmlListDevice> allDevices, final XmlListDevice inputDevice) {
        return new Predicate<XmlListDevice>() {
            @Override
            public boolean apply(XmlListDevice svgDevice) {
                if (svgDevice == null) {
                    return false;
                }
                final Optional<String> logDeviceName = svgDevice.getInternal("LOGDEVICE");
                if (!logDeviceName.isPresent()) {
                    return false;
                }

                Optional<XmlListDevice> logDevice = from(allDevices)
                        .firstMatch(hasName(logDeviceName.get()));
                if (!logDevice.isPresent()) {
                    return false;
                }
                Optional<String> logDeviceRegexp = logDevice.get().getInternal("REGEXP");
                return logDeviceRegexp.isPresent() &&
                        ConcernsDevicePredicate.forPattern(logDeviceRegexp.get()).apply(inputDevice);
            }
        };
    }

    private boolean isConfigDb(ImmutableSet<XmlListDevice> allDevices) {
        return from(allDevices)
                .anyMatch(new Predicate<XmlListDevice>() {
                    @Override
                    public boolean apply(XmlListDevice input) {
                        return input != null && "configDB".equals(input.getAttribute("configfile").orNull());
                    }
                });
    }
}

File: src/main/java/li/klass/fhem/service/device/GraphDefinitionsForDeviceService.kt
package li.klass.fhem.service.device

import android.content.Context
import com.google.common.base.Optional
import com.google.common.collect.ImmutableSet
import li.klass.fhem.domain.log.ConcernsDevicePredicate
import li.klass.fhem.service.graph.gplot.GPlotHolder
import li.klass.fhem.service.graph.gplot.SvgGraphDefinition
import li.klass.fhem.service.room.RoomListService
import li.klass.fhem.service.room.xmllist.XmlListDevice
import org.slf4j.LoggerFactory
import java.util.*
import javax.inject.Inject

class GraphDefinitionsForDeviceService @Inject constructor(
        private val roomListService: RoomListService,
        private val gPlotHolder: GPlotHolder) {

    fun graphDefinitionsFor(context: Context, device: XmlListDevice, connectionId: Optional<String>): Set<SvgGraphDefinition> {
        val allDevices = roomListService.getAllRoomsDeviceList(connectionId, context).allDevicesAsXmllistDevice

        LOGGER.info("graphDefinitionsFor(name={},connection={})", device.name, connectionId.or("--"))
        val graphDefinitions = getGraphDefinitionsFor(allDevices, device, context)
        for (svgGraphDefinition in graphDefinitions) {
            LOGGER.info("graphDefinitionsFor(name={},connection={}) - found SVG with name {}", device.name, connectionId.or("--"), svgGraphDefinition.name)
        }
        return graphDefinitions
    }

    private fun getGraphDefinitionsFor(allDevices: ImmutableSet<XmlListDevice>, device: XmlListDevice, context: Context): Set<SvgGraphDefinition> =
            when (device.type) {
                "SVG" -> setOf(toGraphDefinition(allDevices, context, device))
                else -> allDevices
                        .filter { it.type == "SVG" }
                        .filter { hasConcerningLogDevice(allDevices, device, it) }
                        .filter { gplotDefinitionExists(allDevices, context, it) }
                        .map { toGraphDefinition(allDevices, context, it) }
                        .toSet()
            }

    private fun toGraphDefinition(allDevices: ImmutableSet<XmlListDevice>, context: Context, currentDevice: XmlListDevice): SvgGraphDefinition {
        val logDeviceName = currentDevice.getInternal("LOGDEVICE").get()
        val gplotFileName = currentDevice.getInternal("GPLOTFILE").get()
        val gPlotDefinition = gPlotHolder.definitionFor(gplotFileName, isConfigDb(allDevices), context).get()

        val labels = Arrays.asList(*currentDevice.getAttribute("label")
                .or("").replace("\"".toRegex(), "").split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
        val title = currentDevice.getAttribute("title").or("")
        val plotfunction = Arrays.asList(*currentDevice.getAttribute("plotfunction").or("").trim { it <= ' ' }.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())

        return SvgGraphDefinition(currentDevice.name, gPlotDefinition, logDeviceName, labels, title, plotfunction)
    }

    private fun gplotDefinitionExists(allDevices: ImmutableSet<XmlListDevice>, context: Context, currentDevice: XmlListDevice): Boolean {
        val gplotFileName = currentDevice.getInternal("GPLOTFILE")
        return gplotFileName.isPresent && gPlotHolder.definitionFor(gplotFileName.get(), isConfigDb(allDevices), context).isPresent
    }

    private fun hasConcerningLogDevice(allDevices: ImmutableSet<XmlListDevice>, inputDevice: XmlListDevice, currentDevice: XmlListDevice): Boolean {
        val logDeviceName = currentDevice.getInternal("LOGDEVICE")
        if (!logDeviceName.isPresent) {
            return false
        }

        val logDevice = allDevices.firstOrNull { it.name == logDeviceName.get() }
        logDevice ?: return false

        val logDeviceRegexp = logDevice.getInternal("REGEXP")
        return logDeviceRegexp.isPresent && ConcernsDevicePredicate.forPattern(logDeviceRegexp.get()).apply(inputDevice)
    }

    private fun isConfigDb(allDevices: ImmutableSet<XmlListDevice>): Boolean {
        return allDevices
                .any { it != null && "configDB" == it.getAttribute("configfile").orNull() }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(GraphDefinitionsForDeviceService::class.java)
    }
}
